src/rsync/rsync.c

```diff
@@ -154,21 +154,41 @@ read_pipe(int fd_pipe[2][2], int type)
 {
 	char buffer[4096];
 	ssize_t count;
-	int error;
+	struct timeval timeout;
+	fd_set read_fds;
+	int error, res;
 
 	while (1) {
-		count = read(fd_pipe[type][0], buffer, sizeof(buffer));
-		if (count == -1) {
-			error = errno;
-			if (error == EINTR)
+		timeout.tv_sec = config_get_rsync_transfer_timeout();
+		timeout.tv_usec = 0;
+
+		FD_ZERO(&read_fds);
+		FD_SET(fd_pipe[type][0], &read_fds);
+
+		res = select(fd_pipe[type][0] + 1, &read_fds, NULL, NULL, &timeout);
```

Comment:
One more thing: Why `select(2)` and not `poll(2)`?

[`select(2)`](https://man7.org/linux/man-pages/man2/select.2.html)'s BUGS section mentions malignant quirks abound when `fd ≥ FD_SETSIZE`, and [`pipe(2)`](https://man7.org/linux/man-pages/man2/pipe.2.html) (the current origin of `fd_pipe`) does not constraint itself to any particular range.